packages/checkout-storefront/src/sections/PaymentSection/AdyenDropIn/useAdyenDropin.ts

```diff
@@ -0,0 +1,321 @@
+        setIsProcessingPayment,
+        showCustomErrors,
+        transactionProccess,
+      ]
+    )
+  );
+
+  const onSubmitInitialize: AdyenCheckoutInstanceOnSubmit = useEvent(async (state, component) => {
+    component.setStatus("loading");
+    setAdyenCheckoutSubmitParams({ state, component });
+    validateAllForms(authenticated);
+    setSubmitInProgress(true);
+  });
+
+  useEffect(() => {
+    const validating = anyFormsValidating(validationState);
+    const allFormsValid = areAllFormsValid(validationState);
+
+    // any of the conditions below - do nothing
+    if (!submitInProgress || validating || anyRequestsInProgress || !adyenCheckoutSubmitParams) {
+      return;
+    }
+
+    // there was en error either in some other request or form validation
+    // - stop the submission altogether
+    if (!finishedApiChangesWithNoError || !allFormsValid) {
+      adyenCheckoutSubmitParams?.component.setStatus("ready");
+      setSubmitInProgress(false);
+      return;
+    }
+
+    // submit in progress only means that submit has been initialized
+    // we want to disable it here so it's not run again until the user
+    // initializes manually
+    setSubmitInProgress(false);
+    adyenCheckoutSubmitParams.component.setStatus("loading");
+
+    // there is a previous transaction going on, we want to process instead of initialize
+    if (currentTransactionId) {
+      void onTransactionProccess({
+        data: adyenCheckoutSubmitParams?.state.data,
+        id: currentTransactionId,
+      });
+      return;
+    }
+
+    void onTransactionInitialize({
+      checkoutId,
+      amount: totalPrice.gross.amount,
+      paymentGateway: {
+        id,
+        data: {
+          ...adyenCheckoutSubmitParams.state.data,
+          returnUrl: getUrlForTransactionInitialize()?.newUrl,
+        },
+      },
+    });
+  }, [
+    adyenCheckoutSubmitParams,
+    anyRequestsInProgress,
+    checkoutId,
+    currentTransactionId,
+    finishedApiChangesWithNoError,
+    onTransactionInitialize,
+    onTransactionProccess,
+    submitInProgress,
+    totalPrice.gross.amount,
+    validationState,
+    id,
+    setSubmitInProgress,
+  ]);
+
+  const onAdditionalDetails: AdyenCheckoutInstanceOnAdditionalDetails = useEvent(
+    async (state, component) => {
+      setAdyenCheckoutSubmitParams({ state, component });
+      if (currentTransactionId) {
+        adyenCheckoutSubmitParams?.component?.setStatus("loading");
+        setSubmitInProgress(true);
+      }
+    }
+  );
+
+  // handle when page is opened from previously redirected payment
+  useEffect(() => {
+    const { redirectResult, transaction, processingPayment } = getQueryParams();
+
+    if (!redirectResult || !transaction || !processingPayment) {
+      return;
+    }
+
+    const decodedRedirectData = decodeURI(redirectResult);
+
+    setCurrentTransactionId(transaction);
+
+    clearQueryParams("redirectResult", "resultCode");
+
+    void onTransactionProccess({
```

Comment:
I see a small possibility of a race condition here. When we get back to the page after a redirect, this function is called and if it happens quickly, we end up calling `adyenCheckoutSubmitParams?.component.handleAction(action);` but `adyenCheckoutSubmitParams` might not exist yet.

Is it possible?